File: runinator-command-center-flutter/lib/core/navigation/nav_config.dart
// port of core/navigation/nav-config.ts.

import '../domain/icons.dart';
import 'app_tab.dart';

final List<NavSection> navSections = [
  const NavSection(
    label: 'Workspace',
    items: [
      NavItem(tab: AppTab.dev, label: 'Dev', icon: IconName.debug, desktopOnly: true),
      NavItem(
        tab: AppTab.workflows,
        label: 'Workflows',
        icon: IconName.workflow,
        searchPlaceholder: 'Search workflows',
      ),
      NavItem(tab: AppTab.runs, label: 'Runs', icon: IconName.runs, searchPlaceholder: 'Search runs'),
      NavItem(
        tab: AppTab.providers,
        label: 'Providers',
        icon: IconName.box,
        searchPlaceholder: 'Search providers',
      ),
      NavItem(
        tab: AppTab.replicas,
        label: 'Replicas',
        icon: IconName.list,
        searchPlaceholder: 'Search replicas',
      ),
    ],
  ),
  const NavSection(
    label: 'Inbox',
    items: [
      NavItem(
        tab: AppTab.approvals,
        label: 'Approvals',
        icon: IconName.approve,
        endpoint: 'approvals',
        searchPlaceholder: 'Search approvals',
      ),
      NavItem(
        tab: AppTab.notifications,
        label: 'Notifications',
        icon: IconName.bell,
        endpoint: 'notifications',
        searchPlaceholder: 'Search notifications',
      ),
    ],
  ),
  const NavSection(
    label: 'Data',
    items: [
      NavItem(
        tab: AppTab.artifacts,
        label: 'Artifacts',
        icon: IconName.box,
        endpoint: 'artifacts',
        searchPlaceholder: 'Search artifacts',
      ),
      NavItem(
        tab: AppTab.externalItems,
        label: 'External Items',
        icon: IconName.tag,
        endpoint: 'external_items',
        searchPlaceholder: 'Search external items',
      ),
      NavItem(
        tab: AppTab.events,
        label: 'Events',
        icon: IconName.flag,
        endpoint: 'automation_events',
        searchPlaceholder: 'Search events',
      ),
    ],
  ),
  const NavSection(
    label: 'Other',
    items: [
      NavItem(tab: AppTab.gates, label: 'Gates', icon: IconName.gate, searchPlaceholder: 'Search gates'),
      NavItem(
        tab: AppTab.configs,
        label: 'Configs',
        icon: IconName.settings,
        searchPlaceholder: 'Search configs',
      ),
      NavItem(
        tab: AppTab.secrets,
        label: 'Secrets',
        icon: IconName.key,
        searchPlaceholder: 'Search secrets',
      ),
    ],
  ),
  const NavSection(
    label: 'Organization',
    items: [
      NavItem(tab: AppTab.organization, label: 'Organization', icon: IconName.shield),
      NavItem(tab: AppTab.orgResources, label: 'Resources & Billing', icon: IconName.box),
    ],
  ),
  const NavSection(
    label: 'Admin',
    items: [
      NavItem(tab: AppTab.adminSettings, label: 'Settings', icon: IconName.settings, adminOnly: true),
      NavItem(
        tab: AppTab.permissions,
        label: 'Permissions',
        icon: IconName.shield,
        adminOnly: true,
        searchPlaceholder: 'Search users & teams',
      ),
      NavItem(tab: AppTab.deadLetters, label: 'Dead Letters', icon: IconName.flag, adminOnly: true),
      NavItem(tab: AppTab.auditLog, label: 'Audit Log', icon: IconName.list, adminOnly: true),
    ],
  ),
];

final List<AppTab> tabs = navSections.expand((section) => section.items.map((item) => item.tab)).toList();

final Map<AppTab, NavItem> _navItemByTab = {
  for (final section in navSections)
    for (final item in section.items) item.tab: item,
};

NavItem? navItemForTab(AppTab tab) => _navItemByTab[tab];

String? endpointForTab(AppTab tab) => _navItemByTab[tab]?.endpoint;

bool isResourceTab(AppTab tab) {
  final endpoint = endpointForTab(tab);

  if (endpoint == null) {
    return false;
  }

  return endpoint != 'artifacts' && endpoint != 'notifications';
}

List<NavSection> visibleNavSections({required bool canSeeAdmin, required bool isDesktop}) {
  final sections = navSections
      .map((section) => NavSection(
            label: section.label,
            items: section.items.where((item) => !item.adminOnly || canSeeAdmin).toList(),
          ))
      .where((section) => section.items.isNotEmpty)
      .toList();

  if (isDesktop) {
    return sections;
  }

  return sections
      .map((section) => NavSection(
            label: section.label,
            items: section.items.where((item) => !item.desktopOnly).toList(),
          ))
      .where((section) => section.items.isNotEmpty)
      .toList();
}

/// core/ has no browser localStorage dependency; a concrete web platform adapter
/// (future UI pass) supplies one via [setNavStorageReader]. with none configured
/// (as in a `dart test` run) this behaves like the ts source's try/catch fallback
/// path for unavailable storage.
String? Function(String key)? _storageReader;

void setNavStorageReader(String? Function(String key) reader) {
  _storageReader = reader;
}

AppTab readStoredDefaultTab() {
  final stored = _storageReader?.call('command-center.defaultTab');

  if (stored != null && tabs.any((tab) => tab.wire == stored)) {
    return AppTab.fromWire(stored)!;
  }

  return AppTab.workflows;
}

bool readSidebarCollapsed() => _storageReader?.call('command-center.sidebar.collapsed') == 'true';

File: runinator-command-center/src/core/navigation/nav-config.ts
import type { Action } from "../domain/models";
import type { AppTab, NavItem, NavSection } from "./app";

export const navSections: NavSection[] = [
  {
    label: "Build",
    items: [
      {
        tab: "Workflows",
        label: "Workflows",
        icon: "workflow",
        description: "Create or select a workflow, resolve diagnostics, save, then run it.",
        searchPlaceholder: "Search workflows",
      },
      {
        tab: "Pipelines",
        label: "Pipelines",
        icon: "branch",
        description: "Create a pipeline, add workflows, connect hand-offs, then start a run.",
        searchPlaceholder: "Search pipelines",
      },
      {
        tab: "Functions",
        label: "Functions",
        icon: "box",
        description: "Publish a built package, verify exports, then promote an alias deliberately.",
        searchPlaceholder: "Search function packages",
      },
      {
        tab: "Workspaces",
        label: "Workspaces",
        icon: "folder",
        description: "Inspect saved workspace versions, files, and results shared across workflows.",
        searchPlaceholder: "Search workspace keys",
      },
      {
        tab: "Files",
        label: "Files",
        icon: "folder",
        description: "Upload reusable files and pin immutable revisions from workflow inputs.",
        searchPlaceholder: "Search files",
      },
    ],
  },
  {
    label: "Run & review",
    items: [
      {
        tab: "Runs",
        label: "Workflow Runs",
        icon: "runs",
        description: "Select a run to inspect its timeline, outputs, logs, and recovery actions.",
        searchPlaceholder: "Search runs",
      },
      {
        tab: "PipelineRuns",
        label: "Pipeline Runs",
        icon: "runs",
        description: "Start a pipeline run, then inspect each member attempt and hand-off.",
      },
      {
        tab: "Orchestrations",
        label: "Orchestrations",
        icon: "branch",
        description: "Inspect correlated work or configure the adapter that admits it.",
      },
      {
        tab: "Approvals",
        label: "Approvals",
        icon: "approve",
        description: "Review the request and its run context before approving or rejecting it.",
        endpoint: "approvals",
        searchPlaceholder: "Search approvals",
      },
      {
        tab: "Gates",
        label: "Gates",
        icon: "gate",
        description: "Select a blocking gate, verify its run, and record why you open or close it.",
        searchPlaceholder: "Search gates",
      },
    ],
  },
  {
    label: "Integrations",
    items: [
      {
        tab: "Providers",
        label: "Providers",
        icon: "box",
        description: "Verify registered providers, their actions, and required credential scopes.",
        searchPlaceholder: "Search providers",
      },
      {
        tab: "ExecutionProfiles",
        label: "Execution Profiles",
        icon: "key",
        description: "Configure and monitor file-backed identities materialized for provider actions.",
        requires: "credentials:manage",
        searchPlaceholder: "Search execution profiles",
      },
      {
        tab: "ExternalItems",
        label: "External Items",
        icon: "tag",
        description: "Inspect provider-owned records and their linked workflow context.",
        endpoint: "external_items",
        searchPlaceholder: "Search external items",
      },
      {
        tab: "Events",
        label: "Events",
        icon: "flag",
        description: "Trace automation events by provider, run, node, or message.",
        endpoint: "automation_events",
        searchPlaceholder: "Search events",
      },
      {
        tab: "Notifications",
        label: "Notifications",
        icon: "bell",
        description:
          "Review unread alerts and define policies for failures that require attention.",
        endpoint: "notifications",
        searchPlaceholder: "Search notifications",
      },
    ],
  },
  {
    label: "Operate",
    items: [
      {
        // gated: a console cell can start a workflow run, so this is a privilege rather than a view.
        tab: "Console",
        label: "Console",
        icon: "debug",
        description:
          "Run a command, review its output, and keep long-lived work in a saved session.",
        requires: "console:use",
      },
      {
        tab: "Replicas",
        label: "Replicas",
        icon: "list",
        description:
          "Check runtime health first; use drain or restart only on the selected replica.",
        searchPlaceholder: "Search replicas",
      },
      {
        tab: "Schedules",
        label: "Schedules",
        icon: "clock",
        description: "Create a freeze window with a valid time range and the narrowest safe scope.",
        requires: "schedules:manage",
        searchPlaceholder: "Search freeze windows",
      },
      {
        tab: "Configs",
        label: "Configs",
        icon: "settings",
        description: "Store visible JSON configuration and reuse it through scoped references.",
        requires: "secrets:read",
        searchPlaceholder: "Search configs",
      },
      {
        tab: "Secrets",
        label: "Secrets",
        icon: "key",
        description: "Add or rotate a scoped secret; saved values remain write-only.",
        requires: "secrets:read",
        searchPlaceholder: "Search secrets",
      },
      {
        tab: "Dev",
        label: "Dev",
        icon: "debug",
        description: "Start and inspect the local development stack before testing changes.",
        desktopOnly: true,
      },
    ],
  },
  {
    label: "Organization",
    items: [
      {
        tab: "Organization",
        label: "Organization",
        icon: "shield",
        description: "Manage the active organization, members, roles, and teams.",
      },
      {
        tab: "OrgResources",
        label: "Resources & Billing",
        icon: "box",
        description: "Review limits and usage before changing plan or resource allocations.",
      },
    ],
  },
  {
    label: "Administration",
    items: [
      {
        tab: "AdminSettings",
        label: "Settings",
        icon: "settings",
        description: "Change server settings carefully and validate language/runtime paths first.",
        requires: "credentials:manage",
      },
      {
        tab: "Permissions",
        label: "Permissions",
        icon: "shield",
        description: "Grant the minimum required access, then verify users, teams, and API keys.",
        requires: "members:manage",
        requiresPlatformScope: true,
        searchPlaceholder: "Search users & teams",
      },
      {
        tab: "IngressControl",
        label: "Ingress Control",
        icon: "flag",
        description: "Observe, hold, approve, or drop scoped ingress before it reaches orchestration.",
      },
      {
        tab: "AuditLog",
        label: "Audit Log",
        icon: "list",
        description:
          "Filter by actor, action, or resource to reconstruct an administrative change.",
        requires: "audit:read",
      },
    ],
  },
];

const auxiliaryItems: NavItem[] = [
  {
    tab: "Profile",
    label: "Profile & security",
    icon: "lock",
    description: "Manage your account details, signed-in sessions, and personal API keys.",
  },
];

export const tabs: AppTab[] = [
  ...navSections.flatMap((section) => section.items.map((item) => item.tab)),
  ...auxiliaryItems.map((item) => item.tab),
];

const navItemByTab = new Map(
  [...navSections.flatMap((section) => section.items), ...auxiliaryItems].map(
    (item) => [item.tab, item] as const,
  ),
);

const navSectionByTab = new Map(
  navSections.flatMap((section) => section.items.map((item) => [item.tab, section.label] as const)),
);

export function navItemForTab(tab: AppTab) {
  return navItemByTab.get(tab);
}

export function navSectionForTab(tab: AppTab): string {
  return navSectionByTab.get(tab) ?? "Account";
}

export function endpointForTab(tab: AppTab): string | undefined {
  return navItemByTab.get(tab)?.endpoint;
}

export function isResourceTab(tab: AppTab): boolean {
  const endpoint = endpointForTab(tab);

  if (!endpoint) {
    return false;
  }

  return endpoint !== "notifications";
}

export function visibleNavSections(options: {
  can: (action: Action) => boolean;
  isDesktop: boolean;
  isPlatformScope: boolean;
}): NavSection[] {
  const sections = navSections
    .map((section) => ({
      ...section,
      items: section.items.filter(
        (item) =>
          (!item.requires || options.can(item.requires)) &&
          (!item.requiresPlatformScope || options.isPlatformScope),
      ),
    }))
    .filter((section) => section.items.length > 0);

  if (options.isDesktop) {
    return sections;
  }

  return sections
    .map((section) => ({ ...section, items: section.items.filter((item) => !item.desktopOnly) }))
    .filter((section) => section.items.length > 0);
}

export function readStoredDefaultTab(): AppTab {
  try {
    const stored = localStorage.getItem("command-center.defaultTab");

    if (stored && (tabs as string[]).includes(stored)) {
      return stored as AppTab;
    }
  } catch {
    // storage unavailable.
  }

  return "Workflows";
}

export function readSidebarCollapsed(): boolean {
  try {
    return localStorage.getItem("command-center.sidebar.collapsed") === "true";
  } catch {
    return false;
  }
}
